Mathlib/Data/Fin/Basic.lean

```diff
@@ -48,34 +54,74 @@ lemma Fin.sub_def : ∀ (a b : Fin n), a - b = (Fin.mk ((a + (n - b)) % n) (Nat.
   exact Nat.mod_eq_of_lt a.isLt
 
+@[simp] lemma Fin.mod_eq_val (a : Fin n) : a.val % n = a.val := by
+  simp only [Fin.modn_def, Nat.mod_mod]
+  exact Nat.mod_eq_of_lt a.isLt
+
+theorem Fin.mod_eq_of_lt {a b : Fin n} (h : a < b) : a % b = a := by
+  apply Fin.eq_of_val_eq
+  simp only [Fin.mod_def]
+  rw [Nat.mod_eq_of_lt h, Nat.mod_eq_of_lt a.isLt]
+
+theorem Fin.mod_lt : ∀ (i : Fin n.succ) {m : Fin n.succ}, (0 : Fin n.succ) < m → (i % m) < m
+  | ⟨a, aLt⟩, ⟨m, mLt⟩, hp =>  by
+    have _ : (0 : Nat) < m := Fin.val_zero ▸ hp
+    have _ : a % m < m := Nat.mod_lt _ ‹0 < m›
+    simp only [Fin.mod_def, LT.lt]
+    rw [(Nat.mod_eq_of_lt (Nat.lt_trans ‹a % m < m› mLt) : a % m % n.succ = a % m)]
+    exact Nat.mod_lt _ ‹0 < m›
+
 lemma Fin.add_comm (a b : Fin n) : a + b = b + a := by
-apply Fin.eq_of_val_eq
-simp only [Fin.add_def, Nat.add_comm]
+  apply Fin.eq_of_val_eq
+  simp only [Fin.add_def, Nat.add_comm]
 
 @[simp] lemma Fin.add_zero (a : Fin n.succ) : a + 0 = a := by
-apply Fin.eq_of_val_eq
-simp only [Fin.add_def, Fin.val_zero, Nat.add_zero]
-exact Nat.mod_eq_of_lt a.isLt
+  apply Fin.eq_of_val_eq
+  simp only [Fin.add_def, Fin.val_zero, Nat.add_zero]
+  exact Nat.mod_eq_of_lt a.isLt
 
-@[simp] lemma Fin.zero_add (a : Fin n.succ) : 0 + a = a := by
-rw [Fin.add_comm]
-exact Fin.add_zero a
+@[simp] lemma Fin.zero_add (a : Fin n.succ) : 0 + a = a := (Fin.add_comm _ _) ▸ Fin.add_zero a
+
+lemma Fin.mul_comm (a b : Fin n) : a * b = b * a := by
+  apply Fin.eq_of_val_eq
+  simp only [Fin.mul_def, Nat.mul_comm]
 
 @[simp] lemma Fin.zero_mul (a : Fin n.succ) : 0 * a = 0 := by
-apply Fin.eq_of_val_eq
-simp only [Fin.mul_def, Fin.val_zero, Nat.zero_mul]
-simp [Nat.zero_mod]
+  apply Fin.eq_of_val_eq
+  simp only [Fin.mul_def, Fin.val_zero, Nat.zero_mul]
+  simp [Nat.zero_mod]
 
-lemma Fin.mul_comm (a b : Fin n) : a * b = b * a := by
+@[simp] lemma Fin.mul_zero (a : Fin n.succ) : a * 0 = 0 := by
+  rw [Fin.mul_comm]
+  exact Fin.zero_mul a
+
+@[simp] lemma Fin.one_mul (a : Fin n.succ) : 1 * a = a := by
 apply Fin.eq_of_val_eq
-simp only [Fin.mul_def, Nat.mul_comm]
+simp only [Fin.mul_def]
+cases n with
+| zero => simp only [Nat.eq_zero_of_le_zero (Nat.le_of_succ_le_succ a.isLt)]
+| succ n =>
+  have _ : (1 : Fin n.succ.succ).val = 1 := Nat.mod_eq_of_lt (Nat.succ_le_succ (Nat.succ_le_succ (Nat.zero_le n)))
+  rw [‹(1 : Fin n.succ.succ).val = 1›, Nat.one_mul, Nat.mod_eq_of_lt a.isLt]
+
+@[simp] lemma Fin.mul_one (a : Fin n.succ) : a * 1 = a := by
+  rw [Fin.mul_comm, Fin.one_mul]
 
 lemma Fin.add_assoc (a b c : Fin n) : (a + b) + c = a + (b + c) := by
 apply Fin.eq_of_val_eq
-simp only [Fin.mul_def, Fin.add_def, Nat.mod_add_mod, Nat.add_mod_mod, Nat.add_assoc]
+simp only [Fin.add_def, Nat.mod_add_mod, Nat.add_mod_mod, Nat.add_assoc]
 
-instance : Neg (Fin n) where
-  neg a := ⟨(n - a) % n, Nat.mod_lt _ (lt_of_le_of_lt (Nat.zero_le _) a.isLt)⟩
+lemma Fin.mul_assoc (a b c : Fin n) : (a * b) * c = a * (b * c) := by
+  apply Fin.eq_of_val_eq
```

Comment:
The mathlib proof in Lean 3 seems to be much shorter: https://github.com/leanprover-community/mathlib/blob/3578403f0c5cfbef907d1632ee91cee4b3cc6596/src/data/zmod/basic.lean#L47-L49